Add module PoolSNP

abhilesh opened the pull request:
<!--
# nf-core/modules pull request

Many thanks for contributing to nf-core/modules!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the master branch.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
-->


This PR adds a module for [PoolSNP](https://github.com/capoony/PoolSNP), a heuristic SNP caller for pooled sequencing data. 

The module has takes in a `.mpileup` file and reference `.fasta` as inputs and outputs the following  
- `.vcf.gz`, a gzipped VCF file containing allele counts and frequencies for every position and library
- `*cov-*.txt`, a max coverage file containing the maximum coverage thresholds for all chromosomal arms and libraries in the mpileup file
- `*BS.txt.gz`, Optionally, a bad sites file, containing a list of (variable and invariable) sites that did not pass the SNP calling criteria

## PR checklist

Closes #5953 <!-- If this PR fixes an issue, please link it here! -->

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR.
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [X] `nf-core modules test <MODULE> --profile docker`
    - [X] `nf-core modules test <MODULE> --profile singularity`
    - [X] `nf-core modules test <MODULE> --profile conda`

SPPearce commented on modules/nf-core/poolsnp/main.nf:
```diff
@@ -0,0 +1,71 @@
+    input:
+    tuple val(meta)      , path(mpileup)
+    tuple val(meta2)     , path(reference)
+    val(max_cov)
```
I think this is the only required parameter with no default, so you don't need to expose the others as value channels.

abhilesh replied:
Removed the parameters with default values, also removed the corresponding entries from `meta.yml`, these parameters shouldn't be included in the `meta.yml` right?

Passing the other parameters through `ext.args` as we get an empty `.vcf` file with the default values.

SPPearce replied:
Ok, definitely sensible to pass the other parameters for the test to make an actual output

SPPearce commented:
The tool documentation suggests for the `max-cov` option:
> This cutoff can either be newly calculated by providing the threshold percentile or by providing the full path to a max-coverage file.
So we should try and ensure that option is supported.

abhilesh commented:
> The tool documentation suggests for the `max-cov` option:
> 
> > This cutoff can either be newly calculated by providing the threshold percentile or by providing the full path to a max-coverage file.
> > So we should try and ensure that option is supported.

Any suggestions on how to go about implementing this?

The parameter `max-cov` would need to be either a file type or an integer value, but not both. We can't specify optional inputs for nextflow processes so both the input channels need to be declared and then if a file is passed as the parameter, we use that channel instead of the value channel?

abhilesh commented:
- Added the option to specify `max-cov` as either a float value or a tsv file.
- Assert the presence of only one type of input
- Added tests for both cases when `max-cov` is a value channel vs a path channel

SPPearce reviewed (commented):
Can you add a stub test as well please.

SPPearce reviewed (approved):
Looks good to go now
